include/ghex/mpi/communicator.hpp

```diff
@@ -191,6 +216,133 @@ class communicator
     }
 
     void barrier() { MPI_Barrier(m_comm); }
+
+    // range view over the incoming vectors
+    template<typename T>
+    struct const_view
+    {
+        const T*    _begin;
+        const T*    _end;
+        const T*    begin() const noexcept { return _begin; }
+        const T*    end() const noexcept { return _end; }
+        std::size_t size() const noexcept { return _end - _begin; }
+    };
+
+    // This algorithm is a slight adaptation of Simon Frasch's implementation here:
+    // https://github.com/AdhocMan/arbor/commit/75fa66ab983598c1c7f5e064cde4a062e9d030b5#diff-a5a30d5a5b994ff68317d61e212c8d959c209b2fa4caa3f9cd2e738b9e55ef63
+    template<typename F, typename... Ts>
```

Comment:
This also seems to come from arbor! good. Usually in algorithm implementaiton the functor is specified after the data. Is there a reason why here it is inverted? (no need to change it, just curiosity)

Reply:
What is the purpose of this? In my understanding a distributed for each is simply running the loop on the differetn ranks with the local sizes. Why do you need reductions? I did not find uses for this algorithm in this PR (but it is huge, my computer is collapsing :) )

Reply:
- Regarding the first point: that's a limitation of current c++ where the parameter pack in a function must appear last in order to be able to deduce the other types (here `F`).
- The purpose: it is an alternative way to do an `allgather_v` with multiple vectors at once. The difference to a conventional allgather_v is that we never store all data but rather copy over each rank's data one at a time. For large simulations the local memory might simply be too small to store all global data. For now, it is used in the [creation of the unstructured pattern](https://github.com/boeschf/GHEX/blob/612bf7eb6b59310a100b8a942c911889d076a525/include/ghex/unstructured/pattern.hpp#L188). I will probably also apply this algorithm in the structured pattern in a later PR.